isatty_windows.go

```diff
@@ -4,16 +4,88 @@
 package isatty
 
 import (
+	"strings"
 	"syscall"
+	"unicode/utf16"
 	"unsafe"
 )
 
-var kernel32 = syscall.NewLazyDLL("kernel32.dll")
-var procGetConsoleMode = kernel32.NewProc("GetConsoleMode")
+const (
+	fileNameInfo uintptr = 2
+	fileTypePipe         = 3
+)
+
+var (
+	kernel32                         = syscall.NewLazyDLL("kernel32.dll")
+	procGetConsoleMode               = kernel32.NewProc("GetConsoleMode")
+	procGetFileInformationByHandleEx = kernel32.NewProc("GetFileInformationByHandleEx")
+	procGetFileType                  = kernel32.NewProc("GetFileType")
+)
+
+func init() {
+	// Check if GetFileInformationByHandleEx is available.
+	if procGetFileInformationByHandleEx.Find() != nil {
+		procGetFileInformationByHandleEx = nil
+	}
+}
 
 // IsTerminal return true if the file descriptor is terminal.
 func IsTerminal(fd uintptr) bool {
 	var st uint32
 	r, _, e := syscall.Syscall(procGetConsoleMode.Addr(), 2, fd, uintptr(unsafe.Pointer(&st)), 0)
 	return r != 0 && e == 0
 }
+
+func isCygwinPipeName(name string) bool {
```

Comment:
How about adding a comment, something like this?
`Cygwin/MSYS2 PTY has a name like '\{cygwin,msys}-XXXXXXXXXXXXXXXX-ptyN-{from,to}-master'.`

The format of the name of cygwin pty is very clear in my go-iscygpty code, because I use a regexp:
https://github.com/k-takata/go-iscygpty/blob/master/iscygpty_windows.go#L57
But it not easy to understand from your code, I think.

FYI, I used a similar way in iscygpty.c. and I added the following comment:
https://github.com/k-takata/ptycheck/blob/master/iscygpty.c#L132-L133